File: 1798-max-number-of-k-sum-pairs/1798-max-number-of-k-sum-pairs.java
class Solution {
    public int maxOperations(int[] nums, int k) {
        // Sort the array to use two pointers technique
        Arrays.sort(nums);
        
        int l = 0, r = nums.length - 1; // Initialize two pointers
        int ans = 0; // Variable to count valid pairs
        
        // Iterate while left pointer is less than right pointer
        while (l < r) {
            int s = nums[l] + nums[r]; // Calculate the sum of the two pointer elements
            
            if (s == k) { // If the sum equals k, it's a valid pair
                ++ans; // Increment the count of valid pairs
                ++l;   // Move the left pointer to the right
                --r;   // Move the right pointer to the left
            } else if (s > k) { // If the sum is greater than k, reduce the right pointer
                --r;
            } else { // If the sum is less than k, increase the left pointer
                ++l;
            }
        }
        
        return ans; // Return the total number of valid pairs
    }
}

File: 1798-max-number-of-k-sum-pairs/1798-max-number-of-k-sum-pairs.js
var maxOperations = function(nums, k) {
    const cnt = new Map(); // Map to count occurrences of numbers
    let ans = 0; // Initialize answer to count valid pairs
    
    for (const num of nums) {
        const complement = k - num; // Calculate the complement
        // Check if the complement exists in the map and its count is greater than 0
        if (cnt.get(complement) > 0) {
            cnt.set(complement, cnt.get(complement) - 1); // Decrease count for the complement
            ans++; // Found a valid pair
        } else {
            // If the complement doesn't exist or is 0, store the current number
            cnt.set(num, (cnt.get(num) || 0) + 1); // Increment count for the current number
        }
    }
    return ans; // Return total number of valid pairs
};
